### PR TITLE
feat(plugin-iceberg): Add zorder to rewrite_data_files

### PR DESCRIPTION
**Pull request by mohsaka**
## Description

Add the ability to sort by zorder following iceberg spec.
## Motivation and Context

https://github.com/prestodb/presto/issues/27565
## Impact

Documentation updated.

## Test Plan

Added tests for individual zorder calculation functions as well as a test to make sure that the iceberg file is written in the correct order. 

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.
- [ ] If adding new dependencies, verified they have an [OpenSSF Scorecard](https://securityscorecards.dev/#the-checks) score of 5.0 or higher (or obtained explicit TSC approval for lower scores).

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== NO RELEASE NOTE ==
```

## Summary by Sourcery

Extend Iceberg rewrite_data_files procedure to support configurable file selection and custom split planning.

New Features:
- Add procedure options to rewrite_data_files for minimum input files per partition and file-size-based selection, allowing targeted compaction of small or large files.
- Introduce customizable Iceberg procedure context and split source for rewrite_data_files to control split generation based on procedure options.
- Register Iceberg-specific SQL functions, including z-order related functions, in the Iceberg plugin.

Enhancements:
- Refactor Iceberg metadata to use a common, thread-safe procedure context abstraction shared by distributed procedures.
- Enhance rewrite_data_files to respect per-partition thresholds and delete-file handling when deciding which data files to rewrite.
- Extend IcebergSplitSource with an overload that accepts pre-filtered file tasks and target split size for reuse by specialized split sources.

Documentation:
- Update Iceberg connector documentation to describe the new rewrite_data_files options and behavior.

Tests:
- Add comprehensive rewrite_data_files tests covering min-input-files semantics, partitioned tables, delete files, and combined filters.
- Add tests covering min-file-size-bytes, max-file-size-bytes, and their interaction with min-input-files and filters in rewrite_data_files.

## Summary by Sourcery

Add z-order aware sorting support to Iceberg table data rewrite procedures and refactor distributed procedure analysis to carry procedure-specific context.

New Features:
- Introduce a zorder(varbinary) scalar function in the Iceberg plugin for computing Z-order bytes over ROW-typed inputs.
- Allow Iceberg rewrite_data_files to accept sorted_by specifications using a zorder(...) expression and propagate this into the distributed table data rewrite planning.
- Add parser support to detect zorder usage in sorted_by and pass z-order column context through the distributed procedure analysis pipeline.

Enhancements:
- Extend Iceberg page sink and metadata columns to carry and sort on an internal $z_order column when Z-ordering is requested.
- Refactor distributed procedure analysis to use a generic ProcedureAnalysisContext for table data rewrites, enabling richer per-procedure metadata like z-order columns.

Documentation:
- Update Iceberg connector documentation to describe z-order usage with rewrite_data_files and limitations such as unsupported types for Z-ordering.

Tests:
- Add unit and integration tests for the Iceberg zorder function across supported types, null handling, and multi-dimensional ordering.
- Add rewrite_data_files tests validating Z-order-aware file rewriting, rejection of mixed zorder and regular sorted_by columns, and failures for unsupported types like decimal.

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 1]
```diff
@@ -1298,6 +1298,8 @@ Argument Name         required   type            Description
 ``sorted_by``                    array of        Specify an array of one or more columns to use for sorting. When
                                  strings         performing a rewrite, the specified sorting definition must be
                                                  compatible with the table's own sorting property, if one exists.
+                                                 Supports standard column sorting (e.g., ``'col ASC'``) and
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/iceberg.rst]:
```suggestion
                                                 Supports standard column sorting (for example, ``'col ASC'``) and
```
Avoid Latin abbreviations. See the [GitLab documentation style guide recommended word list entry for e.g.](https://docs.gitlab.com/development/documentation/styleguide/word_list/#eg) for discussion and alternatives.

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 2]
```diff
@@ -1298,6 +1298,8 @@ Argument Name         required   type            Description
...
+                                                 z-order sorting (e.g., ``'zorder(col1, col2)'``).
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/iceberg.rst]:
```suggestion
                                                 z-order sorting (for example, ``'zorder(col1, col2)'``).
```
Avoid Latin abbreviations. See the [GitLab documentation style guide recommended word list entry for e.g.](https://docs.gitlab.com/development/documentation/styleguide/word_list/#eg) for discussion and alternatives.

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the documentation! Nice work. Just a couple of phrasing suggestions.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestRewriteDataFilesProcedure.java [hunk 3]
```diff
@@ -1120,6 +1132,117 @@ public void testRewriteDataFilesWithAllOptionsCombined()
...
+            assertEquals(partkeys.size(), 6, "File should contain all 6 rows");
+
+            // Compute expected z-order for our test data
+            // Input pairs: (1,100), (10,200), (2,300), (20,400), (3,500), (30,600)
+            // Z-order interleaves bits from both columns
+            // For our specific test values, the expected order after z-order sorting is:
+            // (1,100), (10,200), (2,300), (20,400), (3,500), (30,600)
+            // This demonstrates z-order considers both dimensions together
+            List<Long> expectedOrderkeys = ImmutableList.of(1L, 10L, 2L, 20L, 3L, 30L);
+            List<Long> expectedPartkeys = ImmutableList.of(100L, 200L, 300L, 400L, 500L, 600L);
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestRewriteDataFilesProcedure.java:1293]: Already have ordered based checks. Added these exact sequence tests so we can make sure that the functions are working properly.

File: presto-main-base/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 4]
```diff
@@ -303,6 +320,27 @@ private RelationPlan createCallDistributedProcedurePlanForTableDataRewrite(Analy
                 .map(columnToVariableMap::get)
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply presto-main-base/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:324]: Addressed.

File: presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java [hunk 5]
```diff
@@ -130,4 +142,48 @@ public interface FinishCallDistributedProcedure
...
+    {
+        List<String> sortFieldStrings = Collections.emptyList();
+        if (sortOrderIndex.isPresent()) {
+            Object value = arguments[sortOrderIndex.getAsInt()];
+            if (value == null) {
+                sortFieldStrings = Collections.emptyList();
+            }
+            else if (value instanceof List<?>) {
+                sortFieldStrings = Collections.unmodifiableList(((List<?>) value).stream()
+                        .map(String.class::cast)
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java]: Addressed.

File: presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java [hunk 6]
```diff
@@ -130,4 +142,48 @@ public interface FinishCallDistributedProcedure
...
+
+    public static Optional<List<String>> extractZOrderColumns(List<String> sortFieldStrings)
+    {
+        return sortFieldStrings.stream()
+                .map(TableDataRewriteDistributedProcedure::parse)
+                .filter(Optional::isPresent).findFirst()
+                .map(Optional::get);
+    }
+
+    private static Optional<List<String>> parse(String input)
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java:207]: Addressed.

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs) 

Pull branch, local doc build, looks good. Thanks!

File: presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java [hunk 7]
```diff
@@ -130,4 +142,88 @@ public interface FinishCallDistributedProcedure
...
+                                .orElseThrow(() -> new PrestoException(
+                                        INVALID_FUNCTION_ARGUMENT,
+                                        "Invalid zorder(...) expression: " + sortField)));
+            }
+            else if (isZorderExpression(sortField)) {
+                // Input looks like a zorder(...) expression but does not match the expected pattern
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Malformed zorder(...) expression: " + sortField);
+            }
```
Comment (hantangwangd) [inline presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java]:
```suggestion
            if (isZorderExpression(sortField)) {
                if (zorderColumns.isPresent()) {
                    throw new PrestoException(
                            INVALID_FUNCTION_ARGUMENT,
                            "Multiple zorder(...) expressions are not supported in sorted_by");
                }
                zorderColumns = Optional.of(
                        parse(sortField)
                                .orElseThrow(() -> new PrestoException(
                                        INVALID_FUNCTION_ARGUMENT,
                                        "Malformed zorder(...) expression: " + sortField)));
            }
```

nit: can we simplify it this way? It seems a bit simpler and might save on extra match.

**Review by hantangwangd — COMMENTED** [review]
Thanks @mohsaka, the change overall looks good to me! Just one small nit.

Also, should we add a release note for this feature?

File: presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java (hunk 7 above)
Comment (mohsaka) [reply presto-spi/src/main/java/com/facebook/presto/spi/procedure/TableDataRewriteDistributedProcedure.java]: Suggestion taken. Thanks!

**Review by hantangwangd — APPROVED** [review]
